pyop2/codegen/builder.py

```diff
@@ -0,0 +1,816 @@
+            index = Sum(off, Sum(self.layer_index, Product(Literal(numpy.int32(-1)),
+                                                           self.bottom_layer)))
+            bottom = Indexed(bottom.values, (index, ))
+            top = Indexed(top.values, (index, ))
+            return bottom, top
+        return None
+
+    @cached_property
+    def layer_extents(self):
+        if self.iteration_region == ON_BOTTOM:
+            start = Indexed(self._layers_array, (self._layer_index, FixedIndex(0)))
+            end = Sum(Indexed(self._layers_array, (self._layer_index, FixedIndex(0))),
+                      Literal(IntType.type(1)))
+        elif self.iteration_region == ON_TOP:
+            start = Sum(Indexed(self._layers_array, (self._layer_index, FixedIndex(1))),
+                        Literal(IntType.type(-2)))
+            end = Sum(Indexed(self._layers_array, (self._layer_index, FixedIndex(1))),
+                      Literal(IntType.type(-1)))
+        elif self.iteration_region == ON_INTERIOR_FACETS:
+            start = Indexed(self._layers_array, (self._layer_index, FixedIndex(0)))
+            end = Sum(Indexed(self._layers_array, (self._layer_index, FixedIndex(1))),
+                      Literal(IntType.type(-2)))
+        elif self.iteration_region == ALL:
+            start = Indexed(self._layers_array, (self._layer_index, FixedIndex(0)))
+            end = Sum(Indexed(self._layers_array, (self._layer_index, FixedIndex(1))),
+                      Literal(IntType.type(-1)))
+        else:
+            raise ValueError("Unknown iteration region")
+        return (Materialise(PackInst(), start, MultiIndex()),
+                Materialise(PackInst(), end, MultiIndex()))
+
+    @cached_property
+    def _layer_index(self):
+        if self.constant_layers:
+            return FixedIndex(0)
+        if self.subset:
+            # FIXME: Should we instead not compress subset layers array?
+            return self._loop_index
+        else:
+            return self.loop_index
+
+    @cached_property
+    def layer_index(self):
+        if self.extruded:
+            start, end = self.layer_extents
+            return RuntimeIndex(start, end,
+                                LogicalAnd(
+                                    Comparison("<=", Zero((), numpy.int32), start),
+                                    Comparison("<=", start, end)),
+                                name="layer")
+        else:
+            return None
+
+    @property
+    def loop_indices(self):
+        if self.extruded:
+            return (self.loop_index, self.layer_index)
+        else:
+            return (self.loop_index, )
+
+    def add_argument(self, arg):
+        interior_horizontal = self.iteration_region == ON_INTERIOR_FACETS
+        if arg._is_dat:
+            if arg._is_mixed:
+                packs = []
+                for a in arg:
+                    shape = (None, *a.data.shape[1:])
```

Comment:
Yeah, it will be wrong for globals. Do you happen to have a good test case?
I remember facing some difficulties when I tried to implement MixedDat by using other classes, but I can't quite recall what are the issues now. Will have another go.
What do you think of the `flat_shape`?
https://github.com/OP2/PyOP2/blob/kaushik/pyop2/codegen/builder.py#L380